File: src/main/java/com/github/zametki/model/VerificationRecord.java
package com.github.zametki.model;

import com.github.mjdbc.DbMapper;
import com.github.mjdbc.Mapper;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;

import static com.github.zametki.util.DateUtils.optionalInstant;

/**
 *
 */
public class VerificationRecord extends Identifiable<VerificationRecordId> {

    public String hash;
    public UserId userId;
    public VerificationRecordType type;
    public String value;
    public Instant creationDate;
    public Instant verificationDate;

    public VerificationRecord() {
    }

    public VerificationRecord(@NotNull User user, @NotNull VerificationRecordType type, @NotNull String value) {
        this.userId = user.id;
        this.type = type;
        this.value = value;
        this.creationDate = Instant.now();
    }

    @Mapper
    public static final DbMapper<VerificationRecord> MAPPER = r -> {
        VerificationRecord res = new VerificationRecord();
        res.id = new VerificationRecordId(r.getInt("id"));
        res.hash = r.getString("hash");
        res.userId = new UserId(r.getInt("user_id"));
        res.type = VerificationRecordType.fromId(r.getInt("type"));
        res.value = r.getString("value");
        res.creationDate = r.getTimestamp("creation_date").toInstant();
        res.verificationDate = optionalInstant(r.getTimestamp("verification_date"));
        return res;
    };


}

File: src/main/java/com/github/zametki/model/VerificationRecord.kt
package com.github.zametki.model

import com.github.mjdbc.DbMapper
import com.github.mjdbc.Mapper
import com.github.zametki.util.DateUtils.optionalInstant
import java.time.Instant

/**
 *
 */
class VerificationRecord : Identifiable<VerificationRecordId> {

    lateinit var hash: String
    var userId: UserId? = null
    lateinit var type: VerificationRecordType
    lateinit var value: String
    lateinit var creationDate: Instant
    lateinit var verificationDate: Instant

    constructor() {}

    constructor(user: User, type: VerificationRecordType, value: String) {
        this.userId = user.id
        this.type = type
        this.value = value
        this.creationDate = Instant.now()
    }

    companion object {

        @Mapper
        @JvmField
        val MAPPER = DbMapper { r ->
            val res = VerificationRecord()
            res.id = VerificationRecordId(r.getInt("id"))
            res.hash = r.getString("hash")
            res.userId = UserId(r.getInt("user_id"))
            res.type = VerificationRecordType.fromId(r.getInt("type"))
            res.value = r.getString("value")
            res.creationDate = r.getTimestamp("creation_date").toInstant()
            res.verificationDate = optionalInstant(r.getTimestamp("verification_date"))
            res
        }
    }


}
